### ISSUE TITLE
Group Sync Task Being Marked as Failed Causing Massive Sync Delays

### ISSUE THREAD

**Issue by bbusschots-mu** [microsoft/o365-moodle#1705]
In the last release of the plugin the group sync task's behaviour changed. In previous versions having individual courses fail because there were no users with an owner role would not mark the entire scheduled task as failed, now, that is what is happening.

Because we sync our Moodle with our student records systems, it is normal for us to have many courses in Moodle that have no users. We also have automatic syncing of new courses enabled.

In previous versions of the plugin the courses with no owners failed to sync, but the plugin recognised that it had run successfully.

Now, the plugin sees those same errors as the entire sync failing, so Moodle starts to delay the execution of the task because it thinks its broken. This is causing massive delays in Teams sync. ATM the only work-around I can find it to manually clear the timeout Moodle puts on erroring tasks.

Please restore the original behaviour, or advise on a workaround.

**Comment by weilai-irl** [comment]
@bbusschots-mu Hi Bart,

Could you please confirm the version of the plugin and Moodle site please. I cannot reproduce the issue in our environments. We have quite a few courses without teachers in our group create task run, and the task reported the issues correctly, and eventually finished. This was tested in both Moodle 3.9 and 3.10.

Regards,
Lai

**Comment by bbusschots-mu** [comment]
Hi Lai,

Plugin version 3.9.4 (2020071506) on Moodle 3.9.6 (Build: 20210325).

To be clear, the task completes, as it always did, but it reports itself to Moodle as failed, so Moodle adds a timeout to it, throttling it right back to the point that it is only executing once a day or so. I'm not happy attaching a full output on a public forum, but I'd be happy to email you a sample log.

**Comment by bbusschots-mu** [comment]
Hi Lai — following up with direct email now. I can confirm we are running an un-patched version of the plugin directly form this Git repo.

**Comment by weilai-irl** [comment]
@bbusschots-mu Hi Bart,

I have located the issue. Applying the pull request in #1706 should solve the problem. The fix will be included in the next release before the end of the month as well.

The error is in the function to update Teams cache, which was added to the end of the group sync tasks in the last release.

1. A Graph API function call was used to return the full list of all Teams in the organisations. This call was created according to Microsoft documentation and guideline.
2. A separate Graph API call was used on each of the Teams received that doesn't have a record in the cache table in Moodle to get the URL to the Team.

It was the second call that caused problem - some of the Teams returned in the first call were actually groups without Teams, therefore the second call failed with the error `No team found with Group Id`.

The fix to the issue is to add error handling to safely ignore groups without Teams.

Thank you for spotting and reporting the bug.

Regards,
Lai
